### ISSUE TITLE
Evaluation of lazy sequences in prerequisite checkers

### ISSUE THREAD

**Issue by fredZen** [marick/Midje#332]
Hi,

Midje seems to get a border case wrong when it's checking lazyseqs inside a prerequisite. 

See: (tested on midje 1.7.0, clojure 1.7.0, jdk 1.8.0-11)

``` clojure
(ns midjetest.core-test
  (:require [midje.sweet :refer :all]))

(unfinished pre-process)
(unfinished process)

; This case fails
(fact "Midje seems to have a problem with lazyseqs"
  (-> (map pre-process [..metaconstant..])
      process) => ..processed..
  (provided
    (pre-process ..metaconstant..) => ..intermediate..
    (process [..intermediate..]) => ..processed..))

; These 3 cases work as expected
(fact "If I force evaluation of the seq, the problem goes away"
  (-> (map pre-process [..metaconstant..])
      doall
      process) => ..processed..
  (provided
    (pre-process ..metaconstant..) => ..intermediate..
    (process [..intermediate..]) => ..processed..))

(fact "If the mapped function is not mocked the problem also goes away"
  (-> (map identity [..metaconstant..])
      process) => ..processed..
  (provided
    (process [..metaconstant..]) => ..processed..))

(fact "There is also no problem if the checking is not done in a prerequisite"
  (map pre-process [..metaconstant..]) => [..intermediate..]
  (provided
    (pre-process ..metaconstant..) => ..intermediate..))
```

**Comment by marick** [comment]
As far as I know, there's no way around this. The problem is that the result of `map` is an unrealized lazy sequence. `pre-process` will only ever be called when something requires the first value of the lazyseq. In this case, though, the lazyseq is passed to a mocked-out `process`. That function is essentially this:

```clojure
(if (any-match-actual-arguments) 
   (use-the-mock-function-that-matches)
   (constantly nil))
```

Since the lazyseq result of the map does not match the `[..intermediate..]` argument, `process` is made to return `nil`, so lazy argument is never forced to evaluate. 

One thing I'm not sure about is this:

```
FAIL at (t_core.clj:13)
You never said #'process would be called with these arguments:
    (())
```

Given the reasoning above, I'd have expected to see `(nil)`. But it wouldn't be too surprising to see something convert a `nil` to an empty sequence. 

I'll do a little digging, see if there's a way to make the issue slightly more obvious.

Did you look somewhere in the documentation and fail to find an explanation? If so, maybe I should put something in that spot.

**Comment by marick** [comment]
These messages are somewhat better:

```
FAIL at (t_core.clj:13)
You never said #'process would be called with these arguments:
    ["<an unrealized lazy sequence>"]

FAIL at (t_core.clj:13)
These calls were not made the right number of times:
    (pre-process ..metaconstant..) [expected at least once, actually never called]
    (process [..intermediate..]) [expected at least once, actually never called]

FAIL "Midje seems to have a problem with lazyseqs" at (t_core.clj:10)
    Expected: ..processed..
      Actual: "`process` returned this string because it was called with an unexpected argument"
```

I was hoping to do more, but I've run out of time for this issue.

**Comment by fredZen** [comment]
Thank you for you time! Those new messages sure look better me.

What I don't get is why they following tests succeed:

```clojure
(ns ecstatic.lazyseq-test
  (:require [midje.sweet :refer :all]))

(unfinished process)

(let [r (range 1 101)]
  (fact "Midje is able to match an unrealized lazyseq with a plain vector, provided the lazyseq was obtained by mapping a non-mocked function"
    (-> (map identity [..metaconstant..])
        process) => ..processed..
    (provided
      (process [..metaconstant..]) => ..processed..)

    (-> (map inc (range 100))
        process) => ..processed..
    (provided
      (process r) => ..processed..)))

(fact "The arguments to process in the examples above are unrealized"
  (realized? (map identity [..metaconstant..])) => falsey
  (realized? (map inc (range 100))) => falsey)
```

To me it looks like unrealized lazyseqs mysteriously start matching when their (unrealized) contents are produced by a non-mocked function.

If you will explain, I would love to hear how that makes them different.

**Comment by fredZen** [comment]
Oh, and since you asked: the places I looked were [Checkers for collections and strings](https://github.com/marick/Midje/wiki/Checkers-for-collections-and-strings), [Checking sequential collections](https://github.com/marick/Midje/wiki/Checking-sequential-collections) and [Checkers within prerequisites](https://github.com/marick/Midje/wiki/Checkers-within-prerequisites)

**Comment by fredZen** [comment]
The problem looks deeper to me than "can't match unrealized lazyseqs with vectors". Another surprising example:

```clojure
(fact "Here I am realizing the lazyseq so I expect it to match, but it doesn't"
  (-> (map pre-process [..metaconstant..])
      process) => ..processed..
  (provided
    (pre-process ..metaconstant..) => ..intermediate..
    (process (checker [actual]
                ((just [..intermediate..]) (doall actual)))) => ..processed..))

(fact "Actually, realizing the lazyseq seems to be enough to trigger the fail (this fails)"
  (-> (map pre-process [..metaconstant..])
      process) => ..processed..
  (provided
    (pre-process ..metaconstant..) => ..intermediate..
    (process (checker [actual]
                (doall actual)
                true)) => ..processed..))

(fact "The difference from the previous test is that I don't realize the lazyseq. The test succeeds."
  (-> (map pre-process [..metaconstant..])
      process) => ..processed..
  (provided
    (process (checker [actual]
                true)) => ..processed..))
```

**Comment by marick** [comment]
In the first example, the sequence is realized because the argument to `process` (`(map inc (range 100)`) is compared (with structure equality) to `(range 1 101)`. That forces both the sequences. 

Note that the lazyseqs in the last fact are different structures (realized separately), even though they are generated by code that looks the same.

**Comment by marick** [comment]
In the second comment, you're definitely onto something. Here's the example I reduced your first `fact` into, trying out some ideas:

```clojure
(unfinished process)
(defn pre-process [n] 5)

(fact "Comment out the `preprocess` prerequisite below, and this suddenly passes"
  (process (map pre-process [1])) => :done
  (provided
    (pre-process 1) => 5      ; comment this out
    (process [5]) => :done))
```

This is a puzzle. Dunno when I'll be able to really dig into it.

**Comment by fredZen** [comment]
Thank you for the time you took to look into this and confirm it. 

I would really like to see this fixed. I will try and have a look myself, although the code looks rather arcane to me and made me realize I'm a grasshopper with a long way to hop. In case I do find something, would it be more helpful if I worked on the development version rather than the last stable version? Is that version potentially shippable? I don't need it to actually ship, but having a usable snapshot would be of value.

Also, pardon me if this is ridiculous, but I figure the worst I can get is a "no": would you consider selling support on this bug? I don't know that I can afford you, I don't know that billable time is the issue when you say you don't know when you'll be able to look deeper. Just saying I would be ready to part with money to see the bug fixed :-)

**Comment by marick** [comment]
Interesting. My thoughts:

- the development version is potentially shippable.
- my constraint is not money but time.


What might be useful (and free!) is to provide a stripped-down project 
on github that shows what you're trying to do. I have a tentative 
suspicion that the issue is not that Midje doesn't handle particular 
cases around laziness but that there's a more testable (and, 
incidentally, more midje-able) way of expressing what you're trying to 
accomplish.

If you set up that project, I'll take a look. No charge.

Frederic Merizen wrote:

**Comment by fredZen** [comment]
Your question got me thinking and unstuck, thank you.

If you're interested in the longer version, here goes: this is a recurring issue when I start a new development. My top(-ish) level functions often start their life about like this:

```clojure
(->> (find-stuff-to-process)
     (map (comp do-something do-something-else))
     aggregate-results)
```

That's when I would get hit with the problem. Generally, I would scramble for some time, begrudgingly abstain from testing the top level function, test the functions underneath and get on with my life. Typically, the comp'ed functions would quickly get factored into their own named function, so that the untested part ends up looking like this:

```clojure
(->> (find-stuff-to-process)
     (map process-item)
     aggregate-results)
```

No enormous loss, but I lose momentum rediscovering the limitation and navigating around it, and it just grates me the wrong way to leave the code untested (in my experience, untested code tends to invite more untested code to accrete around it).

What's different for me this time around:

- I took the time to investigate the issue. Now I know more precisely what triggers the problem, so I guess I won't get thrown off my track so easily when I meet that one again (should have done that a year ago, really. I always felt like I had more pressing things to attend to :-|)
- You just made me realise that I didn't have to leave map in my top-level function. I was mixing levels of abstraction! Instead, I should do it like this:

```clojure
(defn process-items [items]
  (map process-item items))

(defn main []
  (->> (find-stuff-to-process)
       process-items
       aggregate-results))
```

Now I can test main (for what that's worth, but it's a bit more complicated in the real world), and I can test process-item. The only untestable part that's left is process-items, but since it's only a partial application of map to an opaque function, I guess there's not much to test there :)

Does that make sense to you?

**Comment by fredZen** [comment]
Now trying to apply the insight to my actual code... it's a bit less clear-cut... if I missed something important I'll try to reduce it to a simple example and ask if you will help.

**Comment by marick** [comment]
That last would be good. I'm not 100% sure this is an issue with Midje, but rather something that should be described as "navigating a workflow".

**Comment by fredZen** [comment]
I got carried away, my apologies.

2015-10-26 23:45 GMT+01:00 Brian Marick <notifications@github.com>:

**Comment by marick** [comment]
Should we then close this Issue?

**Comment by fredZen** [comment]
I defer to your decision (this is your project after all). However, since you're asking for my opinion:

IMO, there is still an issue, which led you to write "In the second comment, you're definitely onto something [...] This is a puzzle. Dunno when I'll be able to really dig into it". So, for me, that would argue in favour of keeping the issue open and marked as a bug.

I don't mind very strongly: either way, there's nothing keeping me from submitting a pull request if and when I figure out what's going on (whenever that might be, seeing as I'm currently focusing on restructuring my own code).

**Comment by marick** [comment]
Fair enough. I should probably reduce the puzzle case into something that can be turned into a future-fact. 

I'm going to mark it as an enhancement instead of a bug because I'm not sure it's a bug (rather than an inevitability)

**Comment by phillipm** [comment]
Fixed in https://github.com/marick/Midje/pull/440, which is now available on clojars under [1.9.2-alpha4](https://clojars.org/midje/versions/1.9.2-alpha4)

@fredZen the code in your example should now work.
The issue was pretty subtle, I'll refer the curious to the PR
